drivers/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,309 @@
+        mutex_unlock(&(dev->resp_lock));
+    }
+    else {
+        switch (dev->int_state) {
+            /* A successful mac TX command expects 2 kinds of replies:
+                 * - mac_tx_ok: transmission done, no data received
+                 * - mac_rx <port> <data>: transmission done, some data
+                 *                         received on port <port>. */
+            case RN2XX3_INT_STATE_MAC_TX:
+                /* some data are available */
+                dev->resp_buf[dev->resp_size++] = c;
+                /* if module response ends with 'rx ' after 8 received chars,
+                   the module starts receiving data */
+                if (dev->resp_size == 8 && dev->resp_buf[4] == 'r'
+                    && dev->resp_buf[5] == 'x' && dev->resp_buf[6] == ' ') {
+                    /* next received chars correspond to the port number */
+                    dev->int_state = RN2XX3_INT_STATE_MAC_RX_PORT;
+                }
+                break;
+
+            case RN2XX3_INT_STATE_MAC_RX_PORT:
+                dev->resp_buf[dev->resp_size++] = c;
+                if (c == ' ') {
+                    dev->int_state = RN2XX3_INT_STATE_MAC_RX_MESSAGE;
+                }
+                break;
+
+            case RN2XX3_INT_STATE_MAC_RX_MESSAGE:
+                /* read and convert RX data (received in hex chars) */
+                if (c == ' ') {
+                    dev->resp_buf[dev->resp_size++] = c;
+                    break;
+                }
+                dev->rx_tmp_buf[dev->rx_size % 2] = c;
+
+                /* We convert pairs of hex character to bytes on the fly to
+                   save space in memory */
+                if (dev->rx_size != 0 && dev->rx_size % 2) {
+                    rn2xx3_hex_to_bytes(dev->rx_tmp_buf,
+                                        &dev->rx_buf[(dev->rx_size - 1) / 2]);
+                }
+                dev->rx_size++;
+                break;
+
+            default:
+                dev->resp_buf[dev->resp_size++] = c;
+                break;
+        }
+    }
+}
+
+static void _sleep_timer_cb(void *arg)
+{
+    DEBUG("[rn2xx3] exit sleep\n");
+    rn2xx3_t *dev = (rn2xx3_t *)arg;
+    dev->int_state = RN2XX3_INT_STATE_IDLE;
+}
+
+/*
+ * Driver's "public" functions
+ */
+void rn2xx3_setup(rn2xx3_t *dev, const rn2xx3_params_t *params)
+{
+    assert(dev && (params->uart < UART_NUMOF));
+
+    /* initialize device parameters */
+    memcpy(&dev->p, params, sizeof(rn2xx3_params_t));
+
+    /* initialize pins and perform hardware reset */
+    if (dev->p.pin_reset != GPIO_UNDEF) {
+        gpio_init(dev->p.pin_reset, GPIO_OUT);
+        gpio_set(dev->p.pin_reset);
+    }
+    /* UART is initialized later, since interrupts cannot be handled yet */
+}
+
+int rn2xx3_init(rn2xx3_t *dev)
+{
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_RESET);
+
+    /* initialize buffers and locks*/
+    dev->resp_size = 0;
+    dev->cmd_buf[0] = '\0';
```

Comment:
nit-picky: this is inconsistent with the rest of the code, where you set `dev->cmd_buf[0] = 0;`.